<h1 
  style="color:#207d06; border-bottom: 3px solid #207d06;"
</h1>
<h1 
  style="color:#165b33; border-bottom: 3px solid #165b33;"
</h1>
<h1 
  style="color:#146b3a; border-bottom: 3px solid #146b3a;"
</h1>
<h1 
  style="color:#f8b229; border-bottom: 3px solid #f8b229;"
</h1>
<h1 
  style="color:#ea4630; border-bottom: 3px solid #ea4630;"
</h1>
<h1 
  style="color:#bb2528; border-bottom: 3px solid #bb2528;"
</h1>
<h1 
  style="color:#4974a5; border-bottom: 3px solid #bb2528;"
>
  Day 1: Trebuchet?! 
</h1>


<h2 
  style="color:#207d06; border-bottom: 3px solid #207d06; border-top: 3px solid #207d06;"
>
  Part One
</h2>

--- 

## Day 1: Trebuchet?! 

---

Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.

You've been doing this long enough to know that to restore snow operations, you need to check all fifty stars by December 25th.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

You try to ask why they can't just use a weather machine ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") and hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a trebuchet ("please hold still, we need to strap you in").

As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been amended by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.

The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover. On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:
```
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
```

In this example, the calibration values of these four lines are `12`, `38`, `15`, and `77`. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

To begin, get your puzzle input.

Answer: 
 

In [244]:
import pandas as pd


In [245]:
calibration_document = pd.read_csv('./data/Day1.csv', header=0)

cd = calibration_document.copy()
cd.head()


elf_art
0                fivepqxlpninevh2xxsnsgg63pbvdnqptmg
1               eight8zlctbmsixhrvbpjb84nnmlcqkzrsix
2  hkxqfrqmsixfplbkpkdfzzszjxmdjtdkjlprrvr3gghlrq...
3                             zkjkctxvssix1dqb22five
4                                          4dtlmkfnm

In [246]:
def calibration_values(value):
  cal_nums = []
  [cal_nums.append(i) for i in value if i.isnumeric()]
  answer = int(cal_nums[0]+cal_nums[-1]) if len(cal_nums) !=0 else 0
  return answer
  # return int(cal_nums[0]+cal_nums[-1])

def sum_calibration_values(series):
  cal_vals = []
  [cal_vals.append(calibration_values(num)) for num in series]
  return sum(cal_vals)



In [247]:
test = cd['elf_art'][:5].copy()
sum_calibration_values(test)


196

In [248]:
test = ['1abc2', 'pqr3stu8vwx', 'a1b2c3d4e5f', 'treb7uchet']
sum_calibration_values(test)


142

In [249]:
sum_calibration_values(cd['elf_art'])


53651

Your puzzle answer was `53651`.

The first half of this puzzle is complete! It provides one gold star: *


<h2 
  style="color:#207d06; border-bottom: 3px solid #207d06; border-top: 3px solid #207d06;"
>
  Part Two
</h2>

Your calculation isn't quite right. It looks like some of the digits are actually spelled out with letters: one, two, three, four, five, six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real first and last digit on each line. For example:
```
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
```

In this example, the calibration values are `29`, `83`, `13`, `24`, `42`, `14`, and `76`. Adding these together produces `281`.

What is the sum of all of the calibration values?

Answer:

In [250]:
num_dict = {
  'one': 'o1e',
  'two': 't2o',
  'three': 't3e',
  'four': 'f4r',
  'five': 'f5e',
  'six': 's6x',
  'seven': 's7n',
  'eight': 'e8t',
  'nine': 'n9e',
}


In [251]:
def validate_text_nums(value):
  for i in list(num_dict.keys()):
    if i in value:
      value=value.replace(i, str(num_dict[i]))
    else:
      None
  return value


In [252]:
# Test the example givin in the description:
test = ['two1nine', 'eightwothree', 'abcone2threexyz', 'xtwone3four', '4nineeightseven2', 'zoneight234', '7pqrstsixteen']

# Expected answers are `29`, `83`, `13`, `24`, `42`, `14`, and `76`.
for i in range(len(test)):
  new_val = validate_text_nums(test[i])
  print(f'{test[i]} ==> {new_val} ==> {calibration_values(new_val)}\n')



two1nine ==> t2o1n9e ==> 29

eightwothree ==> e8t2ot3e ==> 83

abcone2threexyz ==> abco1e2t3exyz ==> 13

xtwone3four ==> xt2o1e3f4r ==> 24

4nineeightseven2 ==> 4n9ee8ts7n2 ==> 42

zoneight234 ==> zo1e8t234 ==> 14

7pqrstsixteen ==> 7pqrsts6xteen ==> 76



In [253]:
# Expected sum_calibration_value = 281
test = [validate_text_nums(i) for i in test]
sum_calibration_values(test)


281

In [254]:
# Apply changes and calibrated values to working dataframe:
cd['num_art'] = cd['elf_art'].apply(lambda x: validate_text_nums(x))
cd['calibration_value'] = cd['num_art'].apply(lambda x: calibration_values(x))

cd.head()


elf_art  \
0                fivepqxlpninevh2xxsnsgg63pbvdnqptmg   
1               eight8zlctbmsixhrvbpjb84nnmlcqkzrsix   
2  hkxqfrqmsixfplbkpkdfzzszjxmdjtdkjlprrvr3gghlrq...   
3                             zkjkctxvssix1dqb22five   
4                                          4dtlmkfnm   

                                             num_art  calibration_value  
0                  f5epqxlpn9evh2xxsnsgg63pbvdnqptmg                 53  
1                 e8t8zlctbms6xhrvbpjb84nnmlcqkzrs6x                 86  
2  hkxqfrqms6xfplbkpkdfzzszjxmdjtdkjlprrvr3gghlrq...                 63  
3                              zkjkctxvss6x1dqb22f5e                 65  
4                                          4dtlmkfnm                 44

In [256]:
sum_calibration_values(cd['num_art'])


53894

Your puzzle answer was `53894`:

That's the right answer! You are one gold star closer to restoring snow operations.

Both parts of this puzzle are complete! They provide two gold stars: **

At this point, you should return to your Advent calendar and try another puzzle.

If you still want to see it, you can get your puzzle input.


You have completed Day 1!

<h5 style="text-align:right; padding-right: 10%;">
  <a href="#title">Top Of Page</a>
</h5>